jans-cedarling/cedarling/src/bootstrap_config/decode.rs

```diff
@@ -0,0 +1,310 @@
+
+    /// URI where Cedarling can get JSON file with all required metadata about
+    /// Lock Master, i.e. .well-known/lock-master-configuration.
+    ///
+    /// ***Required*** if `LOCK == Enabled`.
+    #[serde(rename = "CEDARLING_LOCK_MASTER_CONFIGURATION_URI", default)]
+    pub lock_master_configuration_uri: Option<String>,
+
+    /// Controls whether Cedarling should listen for SSE config updates.
+    #[serde(rename = "CEDARLING_DYNAMIC_CONFIGURATION", default)]
+    pub dynamic_configuration: FeatureToggle,
+
+    /// SSA for DCR in a Lock Master deployment. The Cedarling will validate this
+    /// SSA JWT prior to DCR.
+    #[serde(
+        rename = "CEDARLING_LOCK_SSA_JWT",
+        default,
+        deserialize_with = "parse_option_string"
+    )]
+    pub lock_ssa_jwt: Option<String>,
+
+    /// How often to send log messages to Lock Master (0 to turn off trasmission).
+    #[serde(rename = "CEDARLING_AUDIT_LOG_INTERVAL", default)]
+    pub audit_log_interval: u64,
+
+    /// How often to send health messages to Lock Master (0 to turn off transmission).
+    #[serde(rename = "CEDARLING_AUDIT_HEALTH_INTERVAL", default)]
+    pub audit_health_interval: u64,
+
+    /// How often to send telemetry messages to Lock Master (0 to turn off transmission).
+    #[serde(rename = "CEDARLING_AUDIT_TELEMETRY_INTERVAL", default)]
+    pub audit_health_telemetry_interval: u64,
+
+    // Controls whether Cedarling should listen for updates from the Lock Server.
+    #[serde(rename = "CEDARLING_LISTEN_SSE", default)]
+    pub listen_sse: FeatureToggle,
+}
+
+#[derive(Default, Debug, PartialEq, Deserialize)]
+pub enum TrustMode {
+    #[default]
+    Strict,
+    None,
+}
+
+#[derive(Debug, PartialEq, Deserialize, Default)]
+pub enum LogType {
+    #[default]
+    Off,
+    Memory,
+    StdOut,
+    Lock,
+}
+
+#[derive(Debug, PartialEq, Deserialize, Default)]
+pub enum FeatureToggle {
+    #[default]
+    Disabled,
+    Enabled,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum BootstrapDecodingError {
+    #[error("Failed to deserialize Bootstrap config: {0}")]
+    Deserialization(#[from] serde_json::Error),
+    #[error("Missing bootstrap property: `CEDARLING_LOG_TTL`. This property is required if `CEDARLING_LOG_TYPE` is set to Memory.")]
+    MissingLogTTL,
+    #[error("Multiple store options were provided. Make sure you only one of these properties is set: `CEDARLING_POLICY_STORE_URI` or `CEDARLING_LOCAL_POLICY_STORE`")]
+    ConflictingPolicyStores,
+    #[error("No Policy store was provided.")]
+    MissingPolicyStore,
+    #[error(
+        "Unsupported policy store file format for: {0}. Supported formats include: JSON, YAML"
+    )]
+    UnsupportedPolicyStoreFileFormat(String),
+}
+
+impl<'de> Deserialize<'de> for BootstrapConfig {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let raw = BootstrapConfigRaw::deserialize(deserializer)?;
```

Comment:
This creates a very tight coupling between `BootstrapConfig` and the implementation of `BootstrapConfigRaw`. Which will prevent any other possible ways to deserialize into `BootstrapConfig`.

That is easily addressed though - turn this function into a  constructor function for `BootstrapConfig`, something like below. 

Obviously some of the dependent code will change a little.

I tried this change, and it wasn't very difficult.

```rust
impl BootstrapConfig {
    pub fn from_raw_config(raw: &BootstrapConfigRaw) -> Result<Self, SomeSensibleErrorHere> {
v v v
```

Reply:
Let me know if you would like me to push the changes I made on some temporary branch.

Reply:
> will prevent any other possible ways to deserialize into `BootstrapConfig`

You're right. I merged [bd41dbd](https://github.com/JanssenProject/jans/commit/bd41dbd2246931b38972884b8b1a0dc49fc8b2c8) into this branch in [7055f48](https://github.com/JanssenProject/jans/pull/10216/commits/7055f48541a1328a8d4c4d54585ff790eee66e1d).